### PR TITLE
Moving Desktop,& Dotnet runtime providers to separate dll

### PR DESCRIPTION
**Pull request by mayankbansal018**
Using this dll as extension, to load appropriate run-times for session

### CODE REVIEW

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 1]
```diff
@@ -20,19 +21,20 @@ namespace Microsoft.VisualStudio.TestPlatform.Client
...
-    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Host;
-    using Microsoft.VisualStudio.TestPlatform.Common.Hosting;
 
     /// <summary>
     /// Implementation for TestPlatform
     /// </summary>
     public class TestPlatform : ITestPlatform
     {
-        private IFileHelper fileHelper;
+        private readonly TestRuntimeProviderManager testHostProviderManager;
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: nit: rename, maybe ?

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 2]
```diff
@@ -20,19 +21,20 @@ namespace Microsoft.VisualStudio.TestPlatform.Client
...
+
+        private readonly IFileHelper fileHelper;
+
         /// <summary>
         /// Initializes a new instance of the <see cref="TestPlatform"/> class.
         /// </summary>
-        public TestPlatform() : this(new TestEngine(), new FileHelper())
+        public TestPlatform() : this(new TestEngine(), new FileHelper(), TestRuntimeProviderManager.Instance)
         {
             this.testHostProviderManager = TestRuntimeProviderManager.Instance;
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: We don't need this statement anymore.

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs (hunk 1 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: will do this in clean up PR. Have to rename lot of variables, & a few classes, & namespaces as well

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs [hunk 3]
```diff
@@ -58,7 +57,7 @@ public string GetDotnetHostFullPath()
                 }
             }
 
-            string errorMessage = String.Format(Resources.NoDotnetExeFound, dotnetExeName);
+            string errorMessage = String.Empty; //String.Format(Resources.NoDotnetExeFound, dotnetExeName);
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs]: Please explain this change.
If it is required, we should clean it up.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DefaultDataCollectionLauncher.cs [hunk 4]
```diff
@@ -66,7 +66,7 @@ public virtual int LaunchDataCollector(IDictionary<string, string> environmentVa
 
             var argumentsString = string.Join(" ", commandLineArguments);
 
-            this.DataCollectorProcess = this.processHelper.LaunchProcess(dataCollectorProcessPath, argumentsString, processWorkingDirectory, environmentVariables, null, null);
+            this.DataCollectorProcess = this.processHelper.LaunchProcess(dataCollectorProcessPath, argumentsString, processWorkingDirectory, environmentVariables, null, null) as Process;
             return this.DataCollectorProcess.Id;
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/DefaultDataCollectionLauncher.cs]: Null check ?

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs (hunk 3 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs]: I would need to add resource file for this project to accommodate for this resource.

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs [hunk 5]
```diff
@@ -50,33 +47,33 @@ public Process LaunchProcess(string processPath, string arguments, string workin
...
+                    {
+                        // Call WaitForExit again to ensure all streams are flushed
+                        var p = sender as Process;
+                        p.WaitForExit();
+                        exitCallBack(p);
+                    };
                 }
 
-                EqtTrace.Verbose("ProcessHelper: Starting process '{0}' with command line '{1}'", processPath, arguments);
+                // EqtTrace.Verbose("ProcessHelper: Starting process '{0}' with command line '{1}'", processPath, arguments);
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs]: Reason for removing diagnostics ?

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs [hunk 6]
```diff
@@ -104,18 +101,18 @@ public string GetProcessName(int processId)
         }
 
         /// <inheritdoc/>
-        public bool TryGetExitCode(Process process, out int exitCode)
+        public bool TryGetExitCode(Object process, out int exitCode)
         {
-            if (process.HasExited)
+            var proc = process as Process;
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs]: null check here as well

File: src/package/VSIXProject/source.extension.vsixmanifest [hunk 7]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
   <Metadata>
-    <Identity Id="Microsoft.TestPlatform.VSIX.Microsoft.0e51abf6-9f5f-401e-a812-4043b87acabf" Version="$version$" Language="en-US" Publisher="Microsoft" />
+    <Identity Id="Microsoft.TestPlatform.VSIX.Microsoft.0e51abf6-9f5f-401e-a812-4043b87acabf" Version="15.1.0" Language="en-US" Publisher="Microsoft" />
```
Comment (singhsarab) [inline src/package/VSIXProject/source.extension.vsixmanifest]: Is this intentional ?

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/TestEngineTests.cs [hunk 8]
```diff
@@ -137,49 +147,241 @@ public void GetExtensionManagerShouldReturnANonNullInstance()
...
+        [ExtensionUri("executor://CustomTestHost")]
+        [FriendlyName("CustomHost")]
+        private class CustomTestHost : ITestRuntimeProvider
+        {
+            public event EventHandler<HostProviderEventArgs> HostLaunched;
+
+            public event EventHandler<HostProviderEventArgs> HostExited;
+
+            public bool Shared { get; private set; }
+
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/TestEngineTests.cs]: nit: space

File: scripts/build.ps1 [hunk 9]
```diff
@@ -257,6 +257,16 @@ function Publish-Package
         Write-Verbose "Move-Item $coreCLR20PackageDir\$file $coreCLRExtensionsDir -Force"
         Move-Item $coreCLR20PackageDir\$file $coreCLRExtensionsDir -Force
     }
+	
+	# Note Note: If there are some dependencies for the TestHostProvider assemblies, those need to be moved too. 
+    $runtimeproviders = @("Microsoft.TestPlatform.TestHostProvider.dll", "Microsoft.TestPlatform.TestHostProvider.pdb")
```
Comment (codito) [inline scripts/build.ps1]: Do we want to keep the name generic? Should the name say that these are NET runtime providers?

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 10]
```diff
@@ -43,19 +44,24 @@ public TestPlatform() : this(new TestEngine(), new FileHelper())
...
         /// The test engine.
         /// </param>
-        protected TestPlatform(ITestEngine testEngine, IFileHelper filehelper)
+        /// <param name="filehelper">
+        /// The filehelper.
+        /// </param>
+        /// <param name="testHostProviderManager">
+        /// The data.
+        /// </param>
+        internal TestPlatform(ITestEngine testEngine, IFileHelper filehelper, TestRuntimeProviderManager testHostProviderManager)
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: Why not protected? We already use the Testable approach for the unit tests.

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs (hunk 3 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs]: It should be part of this PR please.

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/Interfaces/IDotnetHostHelper.cs [hunk 11]
```diff
@@ -6,7 +6,7 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Helpers.Interfaces
     /// <summary>
     /// Helper class for getting info about dotnet host.
     /// </summary>
-    internal interface IDotnetHostHelper
+    public interface IDotnetHostHelper
```
Comment (codito) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/Interfaces/IDotnetHostHelper.cs]: Why should this be public? I would suggest moving this into TestHostProvider implementation.

We need to discuss how DataCollector can consume this.

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs [hunk 12]
```diff
@@ -25,10 +25,20 @@ public class TestEngine : ITestEngine
     {
         #region Private Fields
 
+        private readonly TestRuntimeProviderManager testHostProviderManager;
         private ITestExtensionManager testExtensionManager;
 
         #endregion
 
+        public TestEngine() : this(TestRuntimeProviderManager.Instance)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs]: Require doc.

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs [hunk 13]
```diff
@@ -25,10 +25,20 @@ public class TestEngine : ITestEngine
...
+        {
+        }
+
+        internal TestEngine(TestRuntimeProviderManager testHostProviderManager)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs]: Can be protected?

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs [hunk 14]
```diff
@@ -109,29 +112,13 @@ public ITestExtensionManager GetExtensionManager()
...
-        /// </summary>
-        /// <param name="runConfiguration">
-        /// The run Configuration.
-        /// </param>
-        /// <returns>
-        /// An instance of the test host manager.
-        /// </returns>
-        public ITestRuntimeProvider GetDefaultTestHostManager(RunConfiguration runConfiguration)
+        /// <inheritdoc/>
+        public ITestRuntimeProvider GetDefaultTestHostManager(string runSettingsXml)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs]: May be this need not be an interface method? It's not getting the `Default` test host manager, it just gets the runtime provider which can understand a runsettings.

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs [hunk 15]
```diff
@@ -1,26 +1,23 @@
...
 
-    using Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Helpers.Interfaces;
-    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Helpers.Interfaces;
 
     /// <summary>
     /// Helper class to deal with process related functionality.
     /// </summary>
-    internal class ProcessHelper : IProcessHelper
+    public class ProcessHelper : IProcessHelper
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs]: Does it help to make this `IProcessHelper<T>`? If the consumer is aware of `Process`, it may as well call the `ProcessHelper<Process>` (may be do not require to do type casts).

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs (hunk 5 above)
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs]: Need to add back in this PR.

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs (hunk 14 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs]: yes, it gets the appropriate runtime provider for current settings, but I let it be in Interface since that would require a lot of test changes as well

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Hosting/DotnetTestHostManagerTests.cs [hunk 16]
```diff
@@ -19,6 +19,7 @@ namespace TestPlatform.CrossPlatEngine.UnitTests.Hosting
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client.Interfaces;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Host;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+    using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Helpers.Interfaces;
```
Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Hosting/DotnetTestHostManagerTests.cs]: These tests should move out to TestPlatformProvider.UnitTests. Kindly enable static analysis on new project.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Hosting/DefaultTestHostManagerTests.cs [hunk 17]
```diff
@@ -9,13 +9,12 @@ namespace TestPlatform.CrossPlatEngine.UnitTests.Hosting
...
     using System.Linq;
     using System.Threading.Tasks;
-
-    using Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Helpers.Interfaces;
     using Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Hosting;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client.Interfaces;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Host;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+    using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Helpers.Interfaces;
```
Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Hosting/DefaultTestHostManagerTests.cs]: Tests need to move out of crossplatengine.

File: src/package/VSIXProject/source.extension.vsixmanifest (hunk 7 above)
Comment (codito) [inline src/package/VSIXProject/source.extension.vsixmanifest]: Requires revert.

File: src/Microsoft.TestPlatform.PlatformAbstractions/netstandard1.0/Process/ProcessHelper.cs [hunk 18]
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Helpers
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netstandard1.0/Process/ProcessHelper.cs]: Not sure if there should be any `Helper` in the namespace. Process abstraction directly belongs to PlatformAbstractions since it is a platform specific concept?

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs (hunk 14 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs]: All the test changes should be in one test class, right? It is a smell if it require changes to multiple entities.

**Review by codito — COMMENTED** [review]
Request few changes.

**Comment by codito** [comment]
New assembly should be added to sign.proj.

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs (hunk 5 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs]: limitation since platform abstraction cannot take dependency on coreutils, which contain eqttrace implementation

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs (hunk 6 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs]: resolved

File: src/package/VSIXProject/source.extension.vsixmanifest (hunk 7 above)
Comment (mayankbansal018) [inline src/package/VSIXProject/source.extension.vsixmanifest]: resolved

File: scripts/build.ps1 (hunk 9 above)
Comment (mayankbansal018) [inline scripts/build.ps1]: clean up

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs (hunk 10 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: resolved

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/Interfaces/IDotnetHostHelper.cs (hunk 11 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/Interfaces/IDotnetHostHelper.cs]: leaving here for now, may be clean up later

File: src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs (hunk 13 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs]: resolved

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs (hunk 15 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/Process/ProcessHelper.cs]: Can be taken in separate PR

File: src/Microsoft.TestPlatform.PlatformAbstractions/netstandard1.0/Process/ProcessHelper.cs (hunk 18 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netstandard1.0/Process/ProcessHelper.cs]: resolved

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs (hunk 3 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs]: resolved

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 19]
```diff
@@ -20,21 +21,21 @@ namespace Microsoft.VisualStudio.TestPlatform.Client
...
-        private IFileHelper fileHelper;
+        private readonly TestRuntimeProviderManager testHostProviderManager;
+
+        private readonly IFileHelper fileHelper;
+
         /// <summary>
         /// Initializes a new instance of the <see cref="TestPlatform"/> class.
         /// </summary>
-        public TestPlatform() : this(new TestEngine(), new FileHelper())
+        public TestPlatform() : this(new TestEngine(), new FileHelper(), TestRuntimeProviderManager.Instance)
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: I'm super confused: Whose responsibility is the TestRuntimeProvider? Not the engine's anymore? How would this look if another engine came in instead of the CrossPlatEngine? Is that even interesting anymore? We might need a doc for this please.

Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: The responsibility to provide TestRuntime, is moved to TestRuntimeProviderManager, which for a given session identifies all available TestRuntime, for instance (Desktop, DotNet), & then returns back a particular instance of Runtime, which would work for current settings.

Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: I see. And is this tied to the test engine in any way? What happens if a substitute engine wants to provide its own host process as a default?(say TAEF) Is that supported?

Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: yes, If Taef wants to be the test runtime(i.e replace testhost), then they can provide their runtime whose job is to understand how to launch TAEF, once that is discovered, Taef can continue to host our crossplat engine to do the communication, & running test work, or they can implement their own communication framework which understands the protocol send across wire.

**Review by codito — APPROVED** [review]
